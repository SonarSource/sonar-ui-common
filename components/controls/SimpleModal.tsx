/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Modal, { ModalProps } from './Modal';

export interface ChildrenProps {
  onCloseClick: (event?: React.SyntheticEvent<HTMLElement>) => void;
  onFormSubmit: (event: React.SyntheticEvent<HTMLFormElement>) => void;
  onSubmitClick: (event?: React.SyntheticEvent<HTMLElement>) => void;
  submitting: boolean;
}

interface Props extends ModalProps {
  children: (props: ChildrenProps) => React.ReactNode;
  header: string;
  onClose: () => void;
  onSubmit: () => void | Promise<void | Response>;
}

interface State {
  submitting: boolean;
}

export default class SimpleModal extends React.Component<Props, State> {
  mounted = false;
  state: State = { submitting: false };

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  stopSubmitting = () => {
    if (this.mounted) {
      this.setState({ submitting: false });
    }
  };

  handleCloseClick = (event?: React.SyntheticEvent<HTMLElement>) => {
    if (event) {
      event.preventDefault();
      event.currentTarget.blur();
    }
    this.props.onClose();
  };

  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.submit();
  };

  handleSubmitClick = (event?: React.SyntheticEvent<HTMLElement>) => {
    if (event) {
      event.preventDefault();
      event.currentTarget.blur();
    }
    this.submit();
  };

  submit = () => {
    const result = this.props.onSubmit();
    if (result) {
      this.setState({ submitting: true });
      result.then(this.stopSubmitting, this.stopSubmitting);
    }
  };

  render() {
    const { children, header, onClose, onSubmit, ...modalProps } = this.props;
    return (
      <Modal contentLabel={header} onRequestClose={onClose} {...modalProps}>
        {children({
          onCloseClick: this.handleCloseClick,
          onFormSubmit: this.handleFormSubmit,
          onSubmitClick: this.handleSubmitClick,
          submitting: this.state.submitting
        })}
      </Modal>
    );
  }
}
