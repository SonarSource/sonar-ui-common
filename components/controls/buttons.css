/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  vertical-align: middle;
  height: var(--controlHeight);
  line-height: calc(var(--controlHeight) - 2px);
  padding: 0 var(--gridSize);
  border: 1px solid var(--darkBlue);
  border-radius: 2px;
  box-sizing: border-box;
  background: transparent;
  color: var(--darkBlue);
  font-weight: 600;
  font-size: var(--smallFontSize);
  text-decoration: none;
  cursor: pointer;
  outline: none;
  transition: border-color 0.2s ease, box-shadow 0.2s ease;
}

.button:hover,
.button.button-active {
  background: var(--darkBlue);
  color: #fff;
}

.button:active {
  box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.125);
}

.button:focus {
  box-shadow: 0 0 0 3px rgba(35, 106, 151, 0.25);
}

.button-primary {
  background: var(--darkBlue);
  color: #fff;
}

.button-primary:hover {
  background: var(--veryDarkBlue);
}

.button.disabled {
  color: var(--disableGrayText) !important;
  border-color: var(--disableGrayBorder) !important;
  background: var(--disableGrayBg) !important;
  cursor: not-allowed !important;
  box-shadow: none !important;
}

/* #region .button-red */
.button-red {
  border-color: var(--red);
  color: var(--red);
}

.button-red:hover,
.button-red.active {
  background: var(--red);
  color: #fff;
}

.button-red:focus {
  box-shadow: 0 0 0 3px rgba(212, 51, 63, 0.25);
}

/* #endregion */

/* #region .button-success */
.button-success {
  border-color: var(--green);
  color: var(--green);
}

.button-success:hover,
.button-success.active {
  background: var(--green);
  color: #fff;
}

.button-success:focus {
  box-shadow: 0 0 0 3px rgba(0, 170, 0, 0.25);
}

/* #endregion */

/* #region .button-link */
.button-link {
  display: inline-flex;
  height: auto;
  /* Keep this to not inherit the height from .button */
  line-height: 1;
  margin: 0;
  padding: 0;
  border: none;
  border-radius: 0;
  background: transparent;
  color: var(--darkBlue);
  border-bottom: 1px solid var(--lightBlue);
  font-weight: 400;
  font-size: inherit;
  transition: border-color 0.2s ease, box-shadow 0.2s ease, color 0.2s ease, border-bottom 0.2s ease;
}

.dropdown .button-link {
  border-bottom: none;
}

.button-link:hover {
  background: transparent;
  color: var(--blue);
}

.button-link:active,
.button-link:focus {
  box-shadow: none;
  outline: 1px dotted var(--blue);
}

.button-link.disabled {
  color: var(--secondFontColor);
  background: transparent !important;
  cursor: default;
}

/* #endregion */

.button-small {
  height: var(--smallControlHeight);
  line-height: 18px;
  padding: 0 6px;
  font-size: 11px;
}

.button-tiny {
  height: var(--tinyControlHeight);
  line-height: var(--tinyControlHeight);
  padding: 0 calc(var(--gridSize) / 2);
}

.button-large {
  height: var(--largeControlHeight);
  padding: 0 16px;
  font-size: var(--mediumFontSize);
}

.button-huge {
  flex-direction: column;
  padding: calc(2 * var(--gridSize));
  width: 200px;
  height: 200px;
  background-color: #fff;
  border: solid 1px #fff;
  border-radius: 3px;
  transition: all 0.2s ease;
  box-shadow: 0 1px 1px 1px var(--barBorderColor);
}

.button-huge:hover,
.button-huge:focus,
.button-huge:active {
  background-color: #fff;
  color: var(--darkBlue);
  box-shadow: var(--defaultShadow);
  transform: translateY(-2px);
}

/* #region .button-group */
/* TODO drop usage of this class in SQ (already dropped from SC) */
.button-group {
  display: inline-block;
  vertical-align: middle;
  font-size: 0;
  white-space: nowrap;
}

.button-group > button,
.button-group > .button {
  position: relative;
  z-index: var(--normalZIndex);
  display: inline-block;
  vertical-align: middle;
  margin: 0;
  cursor: pointer;
}

.button-group > .button:hover:not(.disabled),
.button-group > .button:focus:not(.disabled),
.button-group > .button:active:not(.disabled),
.button-group > .button.active:not(.disabled) {
  z-index: var(--aboveNormalZIndex);
}

.button-group > .button.disabled {
  z-index: var(--belowNormalZIndex);
}

.button-group > .button:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.button-group > .button:not(:last-child):not(.dropdown-toggle) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.button-group > .button + .button {
  margin-left: -1px;
}

.button-group > a:not(.button) {
  vertical-align: middle;
  margin: 0 8px;
  font-size: var(--smallFontSize);
}

/* #endregion */

/* #region .button-icon */
.button-icon {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  vertical-align: middle;
  width: var(--controlHeight);
  height: var(--controlHeight);
  padding: 0;
  border: none;
  color: inherit;
}

.button-icon.button-small {
  width: var(--smallControlHeight);
  height: var(--smallControlHeight);
  padding: 0;
}

.button-icon.button-small svg {
  margin-top: 0;
}

.button-icon.button-tiny {
  width: var(--tinyControlHeight);
  height: var(--tinyControlHeight);
  padding: 0;
}

.button-icon.button-tiny svg {
  margin-top: 0;
}

.button-icon:hover,
.button-icon:focus {
  background-color: currentColor;
}

.button-icon:hover svg,
.button-icon:focus svg {
  color: #fff;
}

/* #endregion */

.button-list {
  display: inline-flex;
  justify-content: space-between;
  height: auto;
  border: 1px solid var(--barBorderColor);
  padding: var(--gridSize);
  margin: calc(var(--gridSize) / 2);
  color: var(--secondFontColor);
  font-weight: normal;
}

.button-list:hover {
  background-color: white;
  border-color: var(--blue);
  color: var(--darkBlue);
}

.no-select {
  user-select: none !important;
}
