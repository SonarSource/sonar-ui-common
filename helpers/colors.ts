/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
