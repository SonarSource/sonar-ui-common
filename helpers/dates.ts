/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
