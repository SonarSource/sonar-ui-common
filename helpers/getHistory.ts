/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createHistory, History } from 'history';
import { useRouterHistory } from 'react-router';

let history: History;

function ensureHistory() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  history = useRouterHistory(createHistory)({
    basename: (window as any).baseUrl,
  });
  return history;
}

export default function getHistory() {
  return history ? history : ensureHistory();
}
