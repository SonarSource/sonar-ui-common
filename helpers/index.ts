/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
