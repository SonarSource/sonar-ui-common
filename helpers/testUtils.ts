/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
