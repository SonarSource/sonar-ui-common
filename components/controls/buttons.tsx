/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import ChevronRightIcon from '../icons/ChevronRightIcon';
import ClearIcon, { ClearIconProps } from '../icons/ClearIcon';
import DeleteIcon from '../icons/DeleteIcon';
import EditIcon from '../icons/EditIcon';
import { IconProps } from '../icons/Icon';
import { ThemeConsumer } from '../theme';
import './buttons.css';
import Tooltip, { TooltipProps } from './Tooltip';

type AllowedButtonAttributes = Pick<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  'className' | 'disabled' | 'id' | 'style' | 'title'
>;

interface ButtonProps extends AllowedButtonAttributes {
  autoFocus?: boolean;
  children?: React.ReactNode;
  innerRef?: (node: HTMLElement | null) => void;
  name?: string;
  onClick?: () => void;
  preventDefault?: boolean;
  stopPropagation?: boolean;
  type?: 'button' | 'submit' | 'reset' | undefined;
}

export class Button extends React.PureComponent<ButtonProps> {
  handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    const { disabled, onClick, preventDefault = true, stopPropagation = false } = this.props;

    event.currentTarget.blur();
    if (preventDefault) {
      event.preventDefault();
    }
    if (stopPropagation) {
      event.stopPropagation();
    }

    if (onClick && !disabled) {
      onClick();
    }
  };

  render() {
    const {
      className,
      disabled,
      innerRef,
      onClick,
      preventDefault,
      stopPropagation,
      type = 'button',
      ...props
    } = this.props;
    return (
      // eslint-disable-next-line react/button-has-type
      <button
        {...props}
        aria-disabled={disabled}
        className={classNames('button', className, { disabled })}
        id={this.props.id}
        onClick={this.handleClick}
        ref={this.props.innerRef}
        type={type}
      />
    );
  }
}

export function ButtonLink({ className, ...props }: ButtonProps) {
  return <Button {...props} className={classNames('button-link', className)} />;
}

export function SubmitButton(props: T.Omit<ButtonProps, 'type'>) {
  // do not prevent default to actually submit a form
  return <Button {...props} preventDefault={false} type="submit" />;
}

export function ResetButtonLink(props: T.Omit<ButtonProps, 'type'>) {
  return <ButtonLink {...props} type="reset" />;
}

export interface ButtonIconProps extends ButtonProps {
  'aria-label'?: string;
  'aria-labelledby'?: string;
  className?: string;
  color?: string;
  onClick?: () => void;
  tooltip?: React.ReactNode;
  tooltipProps?: Partial<TooltipProps>;
}

export function ButtonIcon(props: ButtonIconProps) {
  const { className, color, tooltip, tooltipProps, ...other } = props;
  return (
    <ThemeConsumer>
      {(theme) => (
        <Tooltip mouseEnterDelay={0.4} overlay={tooltip} {...tooltipProps}>
          <Button
            className={classNames(className, 'button-icon')}
            stopPropagation={true}
            style={{ color: color || theme.colors.darkBlue }}
            {...other}
          />
        </Tooltip>
      )}
    </ThemeConsumer>
  );
}

interface ClearButtonProps extends ButtonIconProps {
  className?: string;
  iconProps?: ClearIconProps;
  onClick?: () => void;
}

export function ClearButton({ color, iconProps = {}, ...props }: ClearButtonProps) {
  return (
    <ThemeConsumer>
      {(theme) => (
        <ButtonIcon color={color || theme.colors.gray60} {...props}>
          <ClearIcon {...iconProps} />
        </ButtonIcon>
      )}
    </ThemeConsumer>
  );
}

interface ActionButtonProps extends ButtonIconProps {
  className?: string;
  iconProps?: IconProps;
  onClick?: () => void;
}

export function DeleteButton({ iconProps = {}, ...props }: ActionButtonProps) {
  return (
    <ThemeConsumer>
      {(theme) => (
        <ButtonIcon color={theme.colors.red} {...props}>
          <DeleteIcon {...iconProps} />
        </ButtonIcon>
      )}
    </ThemeConsumer>
  );
}

export function EditButton({ iconProps = {}, ...props }: ActionButtonProps) {
  return (
    <ButtonIcon {...props}>
      <EditIcon {...iconProps} />
    </ButtonIcon>
  );
}

export function ListButton({ className, children, ...props }: ButtonProps) {
  return (
    <Button className={classNames('button-list', className)} {...props}>
      {children}
      <ChevronRightIcon />
    </Button>
  );
}
