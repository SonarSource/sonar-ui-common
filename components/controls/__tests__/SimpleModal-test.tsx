/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { click, waitAndUpdate } from '../../../helpers/testUtils';
import { Button } from '../buttons';
import SimpleModal, { ChildrenProps } from '../SimpleModal';

it('renders', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('closes', () => {
  const onClose = jest.fn();
  const children = ({ onCloseClick }: ChildrenProps) => (
    <Button onClick={onCloseClick}>close</Button>
  );
  const wrapper = shallowRender({ children, onClose });
  click(wrapper.find('Button'));
  expect(onClose).toBeCalled();
});

it('submits', async () => {
  const onSubmit = jest.fn(() => Promise.resolve());
  const children = ({ onSubmitClick, submitting }: ChildrenProps) => (
    <Button disabled={submitting} onClick={onSubmitClick}>
      close
    </Button>
  );
  const wrapper = shallowRender({ children, onSubmit });
  wrapper.instance().mounted = true;
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('Button'));
  expect(onSubmit).toBeCalled();
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

function shallowRender({ children = () => <div />, ...props }: Partial<SimpleModal['props']> = {}) {
  return shallow<SimpleModal>(
    <SimpleModal header="" onClose={jest.fn()} onSubmit={jest.fn()} {...props}>
      {children}
    </SimpleModal>
  );
}
