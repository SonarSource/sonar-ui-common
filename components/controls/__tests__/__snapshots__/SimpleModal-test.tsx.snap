// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Modal
  contentLabel=""
  onRequestClose={[MockFunction]}
>
  <div />
</Modal>
`;

exports[`submits 1`] = `
<Modal
  contentLabel=""
  onRequestClose={[MockFunction]}
>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    close
  </Button>
</Modal>
`;

exports[`submits 2`] = `
<Modal
  contentLabel=""
  onRequestClose={[MockFunction]}
>
  <Button
    disabled={true}
    onClick={[Function]}
  >
    close
  </Button>
</Modal>
`;

exports[`submits 3`] = `
<Modal
  contentLabel=""
  onRequestClose={[MockFunction]}
>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    close
  </Button>
</Modal>
`;
